File: tests/go/ice-test/result/amount_result.go
// Package result provides result leaf nodes for testing.
package result

import (
	"context"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
	"github.com/zjn-zjn/ice/tests/go/ice-test/service"
)

var sendService = service.NewSendService()

// AmountResult grants amount to a user.
type AmountResult struct {
	Key   string  `json:"key" ice:"name:用户ID键,desc:从roam中获取用户ID的键名"`
	Value float64 `json:"value" ice:"name:余额值,desc:要发放的余额数量"`
}

// DoResult implements the LeafResult interface.
func (a *AmountResult) DoResult(ctx context.Context, roam *icecontext.Roam) bool {
	uid := roam.ValueDeep(a.Key).Int()
	if uid == 0 || a.Value <= 0 {
		return false
	}

	res := sendService.SendAmount(uid, a.Value)
	roam.Put("SEND_AMOUNT", res)
	return res
}

// AmountResult2 is another variant of amount granting.
type AmountResult2 struct {
	Key   string  `json:"key" ice:"name:用户ID键,desc:从roam中获取用户ID的键名"`
	Value float64 `json:"value" ice:"name:余额值,desc:要发放的余额数量"`
}

// DoResult implements the LeafResult interface.
func (a *AmountResult2) DoResult(ctx context.Context, roam *icecontext.Roam) bool {
	uid := roam.ValueDeep(a.Key).Int()
	if uid == 0 || a.Value <= 0 {
		return false
	}

	res := sendService.SendAmount(uid, a.Value)
	roam.Put("SEND_AMOUNT", res)
	return res
}

File: tests/python/ice-test/result/amount_result.py
"""Amount result leaf nodes for testing."""

import ice
from ice import Roam
from service.send_service import send_service


@ice.leaf("com.ice.test.result.AmountResult", name="发放余额节点", desc="用于发放余额")
class AmountResult:
    """Grants amount to a user."""
    key: str = ""
    value: float = 0.0
    
    def do_roam_result(self, roam: Roam) -> bool:
        uid_val = roam.get_multi(self.key)
        if uid_val is None or self.value <= 0:
            return False
        
        try:
            uid = int(uid_val)
        except (ValueError, TypeError):
            return False
        
        res = send_service.send_amount(uid, self.value)
        roam.put("SEND_AMOUNT", res)
        return res


@ice.leaf("com.ice.test.result.AmountResult2", name="发放余额节点2", desc="另一个发放余额")
class AmountResult2:
    """Another variant of amount granting."""
    key: str = ""
    value: float = 0.0
    
    def do_roam_result(self, roam: Roam) -> bool:
        uid_val = roam.get_multi(self.key)
        if uid_val is None or self.value <= 0:
            return False
        
        try:
            uid = int(uid_val)
        except (ValueError, TypeError):
            return False
        
        res = send_service.send_amount(uid, self.value)
        roam.put("SEND_AMOUNT", res)
        return res

